### PR TITLE
Guard all uses of basic_random_generator for thread safety

### PR DESCRIPTION
**Pull request by bgbsww**
Hopefully addresses #14254

### CODE REVIEW

**Comment by bgbsww** [comment]
@wwmayer Would you mind giving this a quick review to see if it addresses your note on #14254?

**Comment by bgbsww** [comment]
Ugh, looks like I'm ahead of some compilers.  Will fix.

File: src/Mod/Part/App/TopoShapeExpansion.cpp [hunk 1]
```diff
@@ -4057,6 +4058,36 @@ TopoShape& TopoShape::makeElementChamfer(const TopoShape& shape,
...
+//        std::cerr << edgeCheck.IsClosed3d(TopoDS::Edge(edge)) << std::endl;
+//        std::cerr << edgeCheck.IsSeam(TopoDS::Edge(edge),TopoDS::Face(face)) << std::endl;
+//        std::cerr << edgeCheck.CheckVerticesWithCurve3d(TopoDS::Edge(edge)) << std::endl;
+//        std::cerr << edgeCheck.CheckVerticesWithPCurve(TopoDS::Edge(edge),TopoDS::Face(face)) << std::endl;
+//        std::cerr << edgeCheck.CheckCurve3dWithPCurve(TopoDS::Edge(edge),TopoDS::Face(face)) << std::endl;
+    }
+    // Force the failure here.
+    mkChamfer.Build();
+    if ( ! mkChamfer.IsDone() ) {
+        FC_THROWM(Base::CADKernelError, "Cannot chamfer edge");
```
Comment (FlachyJoe) [inline src/Mod/Part/App/TopoShapeExpansion.cpp:4090]: This part is out of this commit context

Comment (bgbsww) [reply src/Mod/Part/App/TopoShapeExpansion.cpp]: Thank-you.  Inadvertent -a.

**Comment by wwmayer** [comment]
The boost [documentation](https://www.boost.org/doc/libs/1_62_0/libs/uuid/uuid.html#Design%20notes) says that these functions are not thread-safe but they are re-entrant. So, an easier way is to not create static but tmp. instances. Would this be an option?

**Comment by bgbsww** [comment]
It would be, but I saw some concerns in my reading about constantly re-seeding the random generator; both performance and output overlap.  That's why I didn't just make it tmp.  We could go all the way to a singleton pattern, but that seemed overkill.

Totally open to being corrected here.

**Comment by xtemp09** [comment]
Would it be better to use a thread-safe random number generator instead? [^1][^2][^3]	


[^1]: https://www.gnu.org/software/gsl/doc/html/rng.html
[^2]: https://github.com/ilqvya/random#thread-local-random
[^3]: https://github.com/vxst/qrand#usage

**Pull request merged**